pkg/ingress/model_build_listener_rules.go

```diff
@@ -3,40 +3,116 @@ package ingress
+//				}
+//				conditions, err := t.buildRuleConditions(ctx, rule, path, enhancedBackend)
+//				if err != nil {
+//					return errors.Wrapf(err, "ingress: %v", k8s.NamespacedName(ing.Ing))
+//				}
+//				actions, err := t.buildActions(ctx, protocol, ing, enhancedBackend)
+//				if err != nil {
+//					return errors.Wrapf(err, "ingress: %v", k8s.NamespacedName(ing.Ing))
+//				}
+//				tags, err := t.buildListenerRuleTags(ctx, ing)
+//				if err != nil {
+//					return errors.Wrapf(err, "ingress: %v", k8s.NamespacedName(ing.Ing))
+//				}
+//				rules = append(rules, Rule{
+//					Conditions: conditions,
+//					Actions:    actions,
+//					Tags:       tags,
+//				})
+//			}
+//		}
+//	}
+//	optimizedRules, err := t.ruleOptimizer.Optimize(ctx, port, protocol, rules)
+//	if err != nil {
+//		return err
+//	}
+//
+//	priority := int64(1)
+//	for _, rule := range optimizedRules {
+//		ruleResID := fmt.Sprintf("%v:%v", port, priority)
+//		_ = elbv2model.NewListenerRule(t.stack, ruleResID, elbv2model.ListenerRuleSpec{
+//			ListenerARN: lsARN,
+//			Priority:    priority,
+//			Conditions:  rule.Conditions,
+//			Actions:     rule.Actions,
+//			Tags:        rule.Tags,
+//		})
+//		priority += 1
+//	}
+//
+//	return nil
+//}
+
 func (t *defaultModelBuildTask) buildListenerRules(ctx context.Context, lsARN core.StringToken, port int64, protocol elbv2model.Protocol, ingList []ClassifiedIngress) error {
 	if t.sslRedirectConfig != nil && protocol == elbv2model.ProtocolHTTP {
 		return nil
 	}
-
 	var rules []Rule
 	for _, ing := range ingList {
-		for _, rule := range ing.Ing.Spec.Rules {
-			if rule.HTTP == nil {
-				continue
-			}
-			paths, err := t.sortIngressPaths(rule.HTTP.Paths)
+		mergedRules, err := t.mergeRulesPerIngress(ing.Ing.Spec.Rules)
+		if err != nil {
+			return err
+		}
+		for _, mergedRule := range mergedRules {
+			host := mergedRule.Host
+			mergedExactPaths, mergedPrefixPaths, mergedImplementationSpecificPaths, err := t.classifyIngressPathsByType(mergedRule.HTTP.Paths)
 			if err != nil {
 				return err
 			}
-			for _, path := range paths {
-				enhancedBackend, err := t.enhancedBackendBuilder.Build(ctx, ing.Ing, path.Backend,
+			for _, mergedPaths := range [][]networking.HTTPIngressPath{mergedExactPaths, mergedPrefixPaths, mergedImplementationSpecificPaths} {
+				if len(mergedPaths) == 0 {
+					continue
+				}
+				// all the paths in the mergedPaths have the same backend and pathType
+				currPathType := mergedPaths[0].PathType
+				currBackend := mergedRule.HTTP.Paths[0].Backend
+				var paths []string
+				for _, path := range mergedPaths {
+					if path.Path != "" {
+						paths = append(paths, path.Path)
+					}
+				}
+				enhancedBackend, err := t.enhancedBackendBuilder.Build(ctx, ing.Ing, currBackend,
 					WithLoadBackendServices(true, t.backendServices),
 					WithLoadAuthConfig(true))
+				conditions, err := t.buildRuleConditions(ctx, host, paths, enhancedBackend, *currPathType)
```

Comment:
Moving this causes any `err` returned from `enhancedBackendBuilder` to be ignored.